File: Leetcode Challenge/07_July_2020/C++/Week 4/1_Binary Tree Zigzag Level Order Traversal.cpp
//Given a binary tree, return the zigzag level order traversal of its nodes' values. (ie, from left to right, then right to left for the next level and alternate between).
//
//For example:
//Given binary tree [3,9,20,null,null,15,7],
//    3
//   / \
//  9  20
//    /  \
//   15   7
//return its zigzag level order traversal as:
//[
//  [3],
//  [20,9],
//  [15,7]
//]

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        if(root == nullptr) {
            return {};
        }
        vector<vector<int>> ans;
        deque<TreeNode*> q;
        q.push_back(root);
        bool dir = false;
        while(!q.empty()) {
            vector<int> level;
            int tam = q.size();
            while(tam--) {
                TreeNode* node = q.front();
                q.pop_front();
                level.push_back(node->val);
                if(node->left != nullptr) {
                    q.push_back(node->left);
                }
                if(node->right != nullptr) {
                    q.push_back(node->right);
                }
            }
            if(dir) {
                reverse(level.begin(), level.end());
            }
            ans.push_back(level);
            dir = !dir;
        }
        return ans;
    }
};

File: Leetcode Challenge/07_July_2020/Python/Week 4/1_Binary Tree Zigzag Level Order Traversal.py
#Given a binary tree, return the zigzag level order traversal of its nodes' values. (ie, from left to right, then right to left for the next level and alternate between).
#
#For example:
#Given binary tree [3,9,20,null,null,15,7],
#    3
#   / \
#  9  20
#    /  \
#   15   7
#return its zigzag level order traversal as:
#[
#  [3],
#  [20,9],
#  [15,7]
#]

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: TreeNode) -> List[List[int]]:
        if not root:
            return []
        q = deque([root])
        ans = []
        order = 1
        while q:
            level = []
            for _ in range(len(q)):
                node = q.popleft()
                level.append(node.val)
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
            ans.append(level[::order])
            order *= (-1)
        return ans